apps/studio/src/server/modules/collection/collection.router.ts

```diff
@@ -62,28 +63,70 @@ export const collectionRouter = router({
           resourceId: !!parentFolderId ? String(parentFolderId) : null,
         })
 
-        const result = await db
-          .insertInto("Resource")
-          .values({
-            permalink,
-            siteId,
-            type: ResourceType.Collection,
-            title: collectionTitle,
-            parentId: parentFolderId ? String(parentFolderId) : null,
-            state: ResourceState.Published,
-          })
-          .returning(defaultCollectionSelect)
-          .executeTakeFirstOrThrow()
-          .catch((err) => {
-            if (get(err, "code") === PG_ERROR_CODES.uniqueViolation) {
+        const result = await db.transaction().execute(async (tx) => {
+          const user = await db
+            .selectFrom("User")
+            .where("id", "=", ctx.user.id)
+            .selectAll()
+            .executeTakeFirstOrThrow(
+              () => new TRPCError({ code: "BAD_REQUEST" }),
+            )
+
+          if (parentFolderId) {
+            const parentFolder = await db
+              .selectFrom("Resource")
+              .where("Resource.id", "=", String(parentFolderId))
+              .where("Resource.siteId", "=", siteId)
+              .select(["Resource.type", "Resource.id"])
+              .executeTakeFirst()
+
+            if (!parentFolder) {
               throw new TRPCError({
-                code: "CONFLICT",
-                message: "A resource with the same permalink already exists",
+                code: "NOT_FOUND",
+                message: "Parent folder does not exist",
               })
             }
-            throw err
+
+            if (parentFolder.type !== "Folder") {
+              throw new TRPCError({
+                code: "BAD_REQUEST",
+                message:
+                  "Collections can only be created inside other folders or at the root",
+              })
+            }
+          }
+
+          const collection = await tx
+            .insertInto("Resource")
+            .values({
+              permalink,
+              siteId,
+              type: ResourceType.Collection,
+              title: collectionTitle,
+              parentId: parentFolderId ? String(parentFolderId) : null,
+              state: ResourceState.Published,
+            })
+            .returningAll()
+            .executeTakeFirstOrThrow()
+            .catch((err) => {
+              if (get(err, "code") === PG_ERROR_CODES.uniqueViolation) {
+                throw new TRPCError({
+                  code: "CONFLICT",
+                  message: "A resource with the same permalink already exists",
+                })
+              }
+              throw err
+            })
+
+          await logResourceEvent(tx, {
+            eventType: "ResourceCreate",
+            delta: { before: null, after: collection },
+            by: user,
           })
 
+          return collection
```

Comment:
question: return type is a bit different now, should we `pick` the defaultCollectionSelect items?

Reply:
done, i omitted it initially because i thought returning more properties to FE is ok since none of it is sensitive and the increase in size is negligible